hive/src/main/java/org/apache/iceberg/hive/legacy/HiveCatalogWithLegacyReadFallback.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.hive.legacy;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.hive.HiveCatalog;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A {@link HiveCatalog} which uses falls back to using Hive metadata to read tables when Iceberg metadata is not
+ * available. If the table is read through Hive metadata, features like time travel, snapshot isolation and incremental
+ * computation are not supported along with any WRITE operations to either the data or metadata.
+ */
+public class HiveCatalogWithLegacyReadFallback extends HiveCatalog {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HiveCatalogWithLegacyReadFallback.class);
+
+  public HiveCatalogWithLegacyReadFallback(Configuration conf) {
+    super(conf);
+  }
+
+  @Override
+  @SuppressWarnings("CatchBlockLogException")
+  public Table loadTable(TableIdentifier identifier) {
+    // Try to load the table using Iceberg metadata first. If it fails, use Hive metadata
+    try {
+      return super.loadTable(identifier);
```

Comment:
We should make sure if the Iceberg table metadata is corrupted, it does not try to load the legacy Hive table, but should throw an exception

Reply:
I think people who are trying to access a table will be upset if they find out that Iceberg ignores Hive metadata even when its own metadata is corrupted. Here are two alternatives for how to handle this situation:

1) Log an error and then fallback to using Hive metadata.
2) Provide a configuration option that determines whether or not Iceberg should fallback to using Hive metadata. If fallback happens, log a WARNing.

Reply:
Corrupt metadata signals that the table has a problem. It is meant to be read as Iceberg table . This is an exceptional condition for which we should throw an exception

Reply:
After some discussion today, we decided to leave the decision of falling back to the caller (i.e. Dali in LI environment). The catalog now only reads the table using Hive metadata. The caller is expected to check whether the table is a valid Iceberg table or not and invoke the correct catalog. We decided to do this as we also saw the need to support other table formats (i.e. Opal), so this if-else would not be sufficient anyways.